A proposed longer-term rework to SLOAD/SSTORE gas costs (#3404)
ethereum/EIPs: EIPS/eip-3404.md

---
eip: 3404
title: Partial removal of refunds
author: Vitalik Buterin (@vbuterin)
discussions-to: https://ethereum-magicians.org/t/eip-3298-removal-of-refunds/5430
status: Draft
type: Standards Track
category: Core
created: 2021-03-16
---

## Simple Summary

Overhaul SSTORE and SLOAD costs and storage refunds.

## Motivation

This EIP is intended to be an overhaul to storage read/write pricing rules that maintains the benefits of EIPs 2200, 2929, 3298 and 3402, and adds additional benefts, while simplifying the structure of rewards.

## Parameters

> | Parameter | Gas |

Author: ```suggestion
| Parameter | Gas |
----------------------------
```

---
eip: 3404
title: Partial removal of refunds
author: Vitalik Buterin (@vbuterin)
discussions-to: https://ethereum-magicians.org/t/eip-3298-removal-of-refunds/5430
status: Draft
type: Standards Track
category: Core
created: 2021-03-16
---

## Simple Summary

Overhaul SSTORE and SLOAD costs and storage refunds.

> ## Motivation

Reviewer: ```suggestion
## Abstract

Adds tracking of cold state access and state modifications so that gas can be charged separately for cold vs hot reads and for reads vs writes.  Also introduces a mechanism for giving a minor incentive for contracts to leave slots with zero values rather than non-zero values.

## Motivation
```
An abstract would be nice here.  A terse human readable version of the specification, ideally a couple sentences long.

This EIP is intended to be an overhaul to storage read/write pricing rules that maintains the benefits of EIPs 2200, 2929, 3298 and 3402, and adds additional benefts, while simplifying the structure of rewards.

## Parameters

| Parameter | Gas |
----------------------------
| `COLD_SLOAD_COST` | `2000` |
| `COLD_ACCOUNT_ACCESS_COST` | `2500` |
| `STORAGE_SLOT_WRITE_COST` | `2900` |
| `STORAGE_SLOT_FILLING_COST` | `17900` |
| `STORAGE_CLEAR_INCENTIVE_GAS` | `1000` |

## Specification

> For each storage slot and for each account, we currently (as of EIP 2929) already store a transaction-scoped data structure keeping track of whether or not it has been _accessed_. We extend this to two flags : `accessed` and `modified`.

Reviewer: ```suggestion
For each storage slot and for each account, we currently (as of [EIP-2929](./eip-2929.md)) already store a transaction-scoped data structure keeping track of whether or not it has been _accessed_. We extend this to two flags : `accessed` and `modified`.
```

> The "base cost" of `SLOAD` and `SSTORE` are 100. We remove the existing "extra costs" for `SSTORE` as defined in EIP 2929.

Reviewer: ```suggestion
The "base cost" of `SLOAD` and `SSTORE` are 100. We remove the existing "extra costs" for `SSTORE` as defined in [EIP-2929](./eip-2929.md).
```

#### New SLOAD and SSTORE and CALL/CREATE rules

> The "original value" of a storage slot is defined as in [EIP 1283](https://eips.ethereum.org/EIPS/eip-1283): it's the value of the storage slot at the start of executing the current transaction. We add new costs according to the following rules:

Reviewer: ```suggestion
The "original value" of a storage slot is defined as in [EIP 1283](./eip-1283.md): it's the value of the storage slot at the start of executing the current transaction. We add new costs according to the following rules:
```

> * If the storage slot is _not_ `accessed` and it gets `SLOAD`ed or `SSTORE`d, charge an additional `COLD_SLOAD_COST` gas and add that storage slot to the `acccessed` set

Reviewer: ```suggestion
* If the storage slot is _not_ in the `accessed` set and it gets `SLOAD`ed or `SSTORE`d, charge an additional `COLD_SLOAD_COST` gas and add that storage slot to the `acccessed` set
```

> * If the storage slot is _not_ `modified`, and it gets `SSTORE`'d to a value other than its original value, the following effects take place:

Reviewer: ```suggestion
* If the storage slot is _not_ in the `modified` set, and it gets `SSTORE`'d to a value other than its original value, the following effects take place:
```

    * Charge an additional `STORAGE_SLOT_FILLING_COST` if the original value is zero
    * Charge an additional `STORAGE_SLOT_WRITE_COST` if the original value is nonzero
    * Add that storage slot to the `modified` set
 
#### Refunds 
 
At the end of a transaction execution, for each storage slot:

* If `modified = True` but the new value is the same as the original value, refund `STORAGE_SLOT_FILLING_COST` if the original value was zero or `STORAGE_SLOT_WRITE_COST` if the original value was nonzero

> * If `modified = True`, the original value is nonzero, and the new value is zero, refund `CLEAR_INCENTIVE_GAS`

Reviewer: ```suggestion
* If `modified = True`, the original value is nonzero, and the new value is zero, refund `STORAGE_CLEAR_INCENTIVE_GAS`
```

It ensures near-optimally fair gas costs for many key applications:

* Anti-reentrancy locks (typically flipped from 0 to 1 right before a child call begins, and then flipped back to 0 when the child call ends) now cost only the gas for _reading_ that storage slot, and not for writing it
* ERC20 approve-and-send (the "approved value" goes from zero to nonzero when the token transfer is approved, and then back to zero when the token transfer processes) similarly only costs storage reading gas and not storage writing gas
* The `CLEAR_INCENTIVE_GAS` adds an incentive to actually clear storage slots, and not leave a value of 1 just in case you want to use that storage slot later.

## Backwards Compatibility

Refunds are currently only applied _after_ transaction execution, so they cannot affect how much gas is available to any particular call frame during execution. Hence, removing them will not break the ability of any code to execute, though it will render some applications economically nonviable.

## Implementation

TBD

> ## Test case changes

Reviewer: ```suggestion
## Test Cases
```

| `COLD_SLOAD_COST` | `2000` |
| `COLD_ACCOUNT_ACCESS_COST` | `2500` |
| `STORAGE_SLOT_WRITE_COST` | `2900` |
| `STORAGE_SLOT_FILLING_COST` | `17900` |
| `STORAGE_CLEAR_INCENTIVE_GAS` | `1000` |

## Specification

For each storage slot and for each account, we currently (as of EIP 2929) already store a transaction-scoped data structure keeping track of whether or not it has been _accessed_. We extend this to two flags : `accessed` and `modified`.

The "base cost" of `SLOAD` and `SSTORE` are 100. We remove the existing "extra costs" for `SSTORE` as defined in EIP 2929.

#### New SLOAD and SSTORE and CALL/CREATE rules

The "original value" of a storage slot is defined as in [EIP 1283](https://eips.ethereum.org/EIPS/eip-1283): it's the value of the storage slot at the start of executing the current transaction. We add new costs according to the following rules:

* If the storage slot is _not_ `accessed` and it gets `SLOAD`ed or `SSTORE`d, charge an additional `COLD_SLOAD_COST` gas and add that storage slot to the `acccessed` set
* If the storage slot is _not_ `modified`, and it gets `SSTORE`'d to a value other than its original value, the following effects take place:

>     * Charge an additional `STORAGE_SLOT_FILLING_COST` if the original (meaning, at the start of the transaction) value is zero

Author: ```suggestion
    * Charge an additional `STORAGE_SLOT_FILLING_COST` if the original value is zero
```
